src/everwillow/statelib/meta.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+import dataclasses
+import typing as tp
+from functools import partial
+
+import jax.tree_util as jtu
+from jaxtyping import PyTree, PyTreeDef
+
+if tp.TYPE_CHECKING:
+    from everwillow.statelib.state import FrozenChainMap, K, State, V
+
+
+__all__ = ["MergeMeta", "TreeDefMeta"]
+
+
+@partial(
+    jtu.register_dataclass,
+    data_fields=[],
+    meta_fields=["treedef", "keys"],
+)
+@dataclasses.dataclass(frozen=True, slots=True)
```

Comment:
for future ref: we shifted to slots here for memory saving and speed. We might have many instances of this in a code and `__dict__`-approach can hoard memory as managing the dicts requires heavy memory use.